### PR TITLE
Adding support to override fields as a string

### PR DESCRIPTION
**Pull request by s7clarke10**
## Problem

There are sometimes cases where the discovered schema incorrectly assumes a field is a integer or may include data quality issues.

## Proposed changes

Provide the user of tap-s3-csv the option to override a list of fields to be strings. This may be useful where a csv file sometimes includes a data quality error and you still want it to be loaded or where the scanning of the csv file incorrectly assumes a field is a integer.

## Types of changes

This code introduces an optional feature of including a list of fields to override in the json array parameter string_overrides. It does not change current behaviour if the feature is not used. This feature works in the same manner as the date_overrides.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

**Comment by s7clarke10** [comment]
I notice that the ci/cd is failing on the integration tests. There error messages mentioned are similar to what I received when I first attempted to run them on my local copy. I needed to export three environment variables which are used by the Integration Test script call via the Makefile. I'm wondering whether the ci/cd script is not setting these environment variable, hence the integration tests failing.

I.e. I needed to run the following before I could run the integration tests.

export TAP_S3_CSV_ACCESS_KEY_ID=<valid AWS access key id>
export TAP_S3_CSV_SECRET_ACCESS_KEY=<valid AWS secret access key>
export TAP_S3_CSV_BUCKET=<valid bucket which the AWS user / role has access to>

**Comment by Samira-El** [comment]
Hi Steve, the reason the CI is failing is because github actions don't share our secrets with PRs from forks.

Will need to figure out something.

**Comment by Samira-El** [comment]
Will need to introduce  login in the integration tests to skip if the aws s3 credentials are empty.
I don't have capacity to do this, care to make it part of this PR or send a new one?

And btw, thanks for this PR, it does make sense, to build the json schema for the stream, we're taking a guess based on a sample of the data, and sometime that can be wrong.

**Comment by s7clarke10** [comment]
Hi @Samira-El,

Sorry for my delay in responding to you on the CI/CD pipeline.

Because the integration tests are automatically run upon a PR, I’m a little confused how the login would work for a non-interactive workflow. In saying this Github CI/CD is something I haven’t much work on, I have been using more Gitlab.

What I was wondering however was whether we could use `pull_request_target` rather than `pull_request` events. I believe from my reading of this https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/ it will allow the workflow to use the tests from Master rather than the PR. It would mean the CI/CD pipeline runs tests you have vetted and are happy with.

I believe it would be a change to this line here https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/a018d4c0f33e8bc5a85648cf3358b3f9cbcbecca/.github/workflows/ci.yaml#L4 from `pull_request` to `pull_request_target`.

My only thoughts if I understand the article correctly is that you wouldn’t be able to automatically test new test functionality introduced (but will ensure secrets are not compromised from a bad PR). For my new feature, the PR introduces a new unit test for the change, therefore there is a question of how do you introduce new tests if they are bundled with a change.

In summary I believe the pull_request_target was put in place to protect secrets when it runs github actions and probably resolves the security risk but compromises the ability to run new tests introduced in a PR.

Thoughts?

**Comment by msardana94** [comment]
This is a super helpful feature! @Samira-El will you be able to review?

**Comment by s7clarke10** [comment]
Sorry for my delay in responding to you on the CI/CD pipeline.

Because the integration tests are automatically run upon a PR, I’m a little confused how the login would work for a non-interactive workflow. In saying this Github CI/CD is something I haven’t much work on, I have been using more Gitlab.

 

What I was wondering however was whether we could use pull_request_target rather than pull_request events. I believe from my reading of this https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/ it will allow the workflow to use the tests from Master rather than the PR. It would mean it calls tests you have vetted and are happy with.

I believe it would be a change to this line here https://github.com/transferwise/pipelinewise-tap-s3-csv/blob/a018d4c0f33e8bc5a85648cf3358b3f9cbcbecca/.github/workflows/ci.yaml#L4from pull_request to pull_request_target.

 

My only thoughts if I understand the article correctly is that you wouldn’t be able to automatically test new test functionality introduced (but will ensure secrets are not compromised from a bad PR). For my new function, the PR introduces a new unit test for the change, therefore there is a question of how you introduce new tests if they are bundled with a change.

 

In summary I believe the pull_request_target was put in place to protect secrets when it runs github actions and probably resolves the security risk but compromises the ability to run new tests introduced in a PR.

 

Thoughts?
